Add parsing of variables and queries to GraphQL analyzer

ogenstad opened the pull request:
This PR updates the InfrahubGraphQLQueryAnalyzer with better support for parsing variables to the overall query/mutation as well as arguments for each nested query within the GraphQL document.

The final goal is to be able to figure out if the query can uniquely identify a single result and then use this information to determine what artifacts etc need to be within the pipeline. It's related to #5540. With what we have in place in this PR the overall situation should be a lot better specifically for instances where we don't add an extra query like in the linked issue.

I'd like to get this merged with the current feature set. But there are a few follow ups that need to be in place before we can close the target issue:

* HFID support (if the query is setup to require an hfid we should be able to say that it would yield a single response
* Better overall support for the uniqueness constraints, which will be relevant if there are multiple constraints on the target node (this might be an extra future feature)
* With the example in #5540 we can clearly see that `InfraServer` is problematic with regards to a single identifying object based on the input. We could still extract the modules that are being queried for both of these subqueries to indicate which ones are within the single response part along with the ones where we'd have to look at any updated object of the same kind
* We need to consider what information should be stored to convey this information within the `CoreGraphqlQuery` node in the database

Another update in this PR was that I wanted to try a new approach to how we setup the test requirements. I introduced a new fixture `test_kit` that returns a TestKit object. The idea here is that we'd be able to simplify the test setup in a way that I think is cleared than what we typically have with loading multiple fixtures that can be verbose in some places and you need to track down what they are doing. This is similar to another fixture `helper`, which we haven't used much of late. The idea here was have a class with a better structure with shortcuts for various things we want to do such as `test_kit.schema` for schema related operations and `test_kit.graphql` for GraphQL related tasks. Then we can add others as needed. At the moment I send the registry to both schema and graphql, it's not used right now in the graphql part, but we should probably refactor that moving. forward so that the prepare_graphql function doesn't use the registry directly.

ogenstad committed:
Add parsing of variables and queries to GraphQL analyzer

ajtmccarty commented on backend/infrahub/graphql/analyzer.py:
```diff
@@ -298,6 +354,22 @@ def kind_action_map(self) -> dict[str, set[MutateAction]]:
         return access
 
+    @property
+    def targets_single_result(self) -> bool:
```
could this be inverted so that it returns `False` early and avoids iterating over the rest of `self.queries` if it has already failed?

ajtmccarty commented on backend/infrahub/graphql/analyzer.py:
```diff
@@ -298,6 +354,22 @@ def kind_action_map(self) -> dict[str, set[MutateAction]]:
+            targets_single_query = False
+            if query.infrahub_model and query.infrahub_model.uniqueness_constraints:
+                for argument in query.arguments:
+                    if [[argument.name]] == query.infrahub_model.uniqueness_constraints:
```
should this be an `in` conditional instead of an `==`?
for example, if I have a schema that is unique on both `name` and `color`, the uniqueness_constraints would be `[["name"],["color"]]`

ogenstad replied:
No I think for now we'll use an exact match. Given how this will be used I think it will be easier if we start off with an exact match for the uniqueness constraint to ensure that we meet the easiest requirement first and then we can expand upon that for more advanced constraints as the next step.

ajtmccarty commented on backend/tests/unit/graphql/test_query_analyzer.py:
```diff
@@ -358,3 +392,206 @@ async def test_query_report(db: InfrahubDatabase, default_branch: Branch, car_pe
+            }
+        }
+    }
+    """
```
maybe I don't understand exactly what "single result" means in this context, but it seems like this query would return between 0 and len(name_values) results. so it seems like this would be a single result query

ogenstad replied:
I think the problem here is that `targets_single_result` is a bad name. What it's ment to convey is if we given the input (hard coded or by using variables) would be able to capture the relevant data to correctly populate a CoreGraphQLQueryGroup.

If we look at the query below we know that we'd be able to correctly add the the required members and ensure that future updates would be correct. As the `$device` variable is a required string and the InfraDevice has a uniqueness constraint on `name__value` we will capture the device along with any current interfaces. If new interfaces are added to the device the device it self would have been updated so so on the next update that new interface would be added as a member to the GraphQL query group

```graphql
query device_config($device: String!) {
  InfraDevice(name__value: $device) {
    edges {
      node {
	name {
	  value
	}
	interfaces {
	  edges {
	    node {
	      name {
		value
	      }
	    }
	  }
	}
      }
    }
  }
}
```

If we look at another example where this is not the case we can look at this query:

```graphql
query device_config($device: String!) {
  InfraDevice(name__value: $device) {
    edges {
      node {
	name {
	  value
	}
	interfaces {
	  edges {
	    node {
	      name {
		value
	      }
	    }
	  }
	}
      }
    }
  }
  InfraServer {
    edges {
      node {
	name {
	  value
	}
      }
    }
  }
}
```

When we execute this query the first time any current `InfraServer` nodes would be added to the GraphQLQuery group, however if you add a new server later that server would not be a member of the group which means that we wouldn't trigger a job (such as the creation of an artifact) when we add a new server.

Because of this the `targets_single_result` property is meant to return `True` for the first query but `False` for the second one. Might "captures_unique_members" or similar be better?

ajtmccarty replied:
thanks for the examples, I think I get it now
maybe `only_has_unique_targets`? this one is hard to name descriptively

ogenstad replied:
Yeah that sounds closer to what we want. I think I'll go with that for now. At the moment it's only an internal fields so I don't know if we need to expose the name in any way.

ajtmccarty commented on backend/tests/unit/graphql/test_query_analyzer.py:
```diff
@@ -358,3 +392,206 @@ async def test_query_report(db: InfrahubDatabase, default_branch: Branch, car_pe
+            }
+        }
+    }
+    """
```
and this one seems like it could be up to 2 results: one for the TShirt and one for the Color

ogenstad commented:
I've removed the TestKit class and updated the previously confusing name to indicate that the query itself should produce a single target instead of a single response.
